ui-tests/viewAllCategories.spec.js

```diff
@@ -0,0 +1,246 @@
+  ).toBeVisible();
+
+  // Click on "Product details button"
+  await page.locator(".btn").first().click();
+
+  // Expect product details to be present on  product details page
+  await expect(
+    page.getByRole("heading", { name: "Product Details" })
+  ).toBeVisible();
+  await expect(
+    page.getByRole("heading", { name: "Name : Wireless mouse with cable" })
+  ).toBeVisible();
+  await expect(
+    page.getByRole("heading", {
+      name: "Description : High-precision and good bluetooth mouse",
+    })
+  ).toBeVisible();
+  await expect(page.getByRole("main")).toContainText("Price :$10");
+  await expect(
+    page.getByRole("heading", {
+      name: "Category : IT accessories and other electronics",
+    })
+  ).toBeVisible();
+  await expect(
+    page.getByRole("img", { name: "Wireless mouse with cable" })
+  ).toBeVisible();
+  await expect(page.getByRole("button", { name: "ADD TO CART" })).toBeVisible();
+
+  // Expect related products to be visible
+  await expect(
+    page.getByRole("heading", { name: "Similar Products ➡️" })
+  ).toBeVisible();
+
+  // Expect the relevant information of 1 related product to be visible
+  await expect(
+    page.getByRole("heading", { name: "Mechanical keyboard with RGB light" })
+  ).toBeVisible();
+  await expect(page.getByRole("heading", { name: "$91.00" })).toBeVisible();
+  await expect(page.getByText("RGB mechanical keyboard")).toBeVisible();
+  await expect(
+    page.getByRole("button", { name: "More Details" }).first()
+  ).toBeVisible();
+
+  // Expect the remaining related products to be visible
+  await expect(
+    page.getByRole("heading", {
+      name: "Electronic toy car",
+    })
+  ).toBeVisible();
+  await expect(
+    page.getByRole("heading", {
+      name: "Wireless bluetooth speaker with charger",
+    })
+  ).toBeVisible();
+}
+
+test.describe("Successful view all categories and related product details", () => {
+  // Check that an unauthorised user can view all categories and related product details
+  test("Should allow unauthorised user to view all categories and related product details", async ({
+    page,
+  }) => {
+    await page.goto("http://localhost:3000", { waitUntil: "commit" });
+    await viewAllCategoryAndRelatedProducts(page);
+  });
+
+  // Check that an authorised user can  view all categories and related product details
+  test("Should allow authorised user to view all categories and related product details", async ({
+    page,
+  }) => {
+    await page.goto("http://localhost:3000", { waitUntil: "commit" });
+
+    await page.getByRole("link", { name: "Login" }).click();
+
+    await page
+      .getByRole("textbox", { name: "Enter Your Email" })
+      .fill("sammyrae@gmail.com");
+    await page
+      .getByRole("textbox", { name: "Enter Your Password" })
+      .fill("srae123");
+    await page.getByRole("button", { name: "LOGIN" }).click();
+
+    await viewAllCategoryAndRelatedProducts(page);
+  });
+});
```

Comment:
another optional comment (or test case identified suggestion):

If we have an object from a certain category, and the rest of the objects are not related, then we dont show any objects as related objects